P0009/P0009r12.md

```diff
@@ -708,89 +709,180 @@ properties will work.
-with parts of the application written in one language and others in a different language. 
-In those cases data is handed over via simple pointers, with the appropriate shaping as 
-multidimensional arrays being superimposed. 
-
-<b>Mixing Compile-Time and Run-Time Dimensions </b>
-
-The fundamental reason to allow for compile time dimensions is performance.
+
+## Why do we need multidimensional arrays?
+
+Multidimensional arrays are fundamental concepts in many fields,
+including graphics, mathematics, statistics, engineering, and the sciences.
+Many programming languages thus come with multidimensional array data structures
+either as a core language feature,
+or as a tightly integrated standard library.
+Example languages include
+Ada, ANSI Common Lisp, APL, C#, Fortran, Julia, Matlab, Mathematica,
+Pascal, Python (via NumPy), and Visual Basic.
+The original version of the Fortran language for the IBM 704
+featured arrays with one, two, or three dimensions (Backus 1956, pp. 10-11).
+
+Multidimensional arrays have long been useful
+for representing large amounts of data,
+describing points in physical space,
+or expressing approximations of functions.
+They are a natural way to represent mathematical objects like matrices and tensors.
+This makes multidimensional arrays a critical data structure for many computations
+at the heart of modern machine learning.
+In fact, one of the predominant machine learning frameworks
+is called [TensorFlow](https://www.tensorflow.org/).
+
+## Why are existing C++ data structures not enough?
+
+C++ currently has the following approaches
+that could be used to represent multidimensional arrays:
+
+1. "native" arrays where all the dimensions are compile-time constants,
+  like `int[3][4][5]`;
+
+2. pointer-of-pointers(-of-pointers...), like `int***`,
+  set up as a data structure to view multidimensional data;
+
+3. arrays-of-arrays(-of-arrays...) data structures,
+  like `vector<vector<array<int, N>>>`; or
+
+4. `gslice`, which selects a subset of indices of a `valarray`
+  and can be used to impose a multidimensional array layout
+  on the `valarray`, in a way analogous to `layout_stride`.
+
+If a multidimensional array has any dimensions
+that are not known at compile time, Approach (1) does not work.
+
+Approach (2) does not suffice as a stand-alone data structure,
+because a pointer-of-pointers does not carry along
+the array's run-time dimensions.
+Users thus end up building some subset of `mdspan`'s functionality
+to represent a multidimensional array view.
+A pointer-of-pointers also loses information
+about any dimensions known at compile time.
+Users cannot arbitrarily mix compile-time and run-time dimensions.
+
+Approach (3) can mix `vector` and `array` to represent dimensions
+known at run time resp. compile time.
+However, a `vector` of `vector`s (of `vector`s) is a container
+with many separate allocations.
+It cannot represent a view of contiguous memory.
+
```

Comment:
Add: both (3) and (4) also have significant performance issues, since for each access to a value multiple memory locations must be read, i.e. pointer chasing occurs.